merlin/models/tf/blocks/multi_optimizers.py

```diff
@@ -0,0 +1,214 @@
+#
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import collections
+from typing import List, Optional, Sequence, Tuple, Union
+
+import tensorflow as tf
+
+import merlin.models.tf as ml
+from merlin.models.tf.core.base import Block
+from merlin.models.tf.utils import tf_utils
+
+Tensor = Union[tf.Tensor, tf.SparseTensor, tf.RaggedTensor]
+
+
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class MultiOptimizer(tf.keras.optimizers.Optimizer):
+    """An optimizer that composes multiple individual optimizers.
+
+    It allows different optimizers to be applied to different subsets of the
+    model's variables. For example, it makes it possible to apply one
+    optimizer to the blocks which contains the model's embeddings (sparse variables)
+    and another optimizer to the rest of its variables (other blocks).
+
+    To specify which optimizer should apply to each block, pass a list of
+    pairs of (optimizer instance, blocks the optimizer should apply to).
+
+    For example:
+    ```python
+      user_tower = ml.InputBlock(schema.select_by_tag(Tags.USER), ml.MLPBlock([512, 256]))
+      item_tower = ml.InputBlock(schema.select_by_tag(Tags.ITEM), ml.MLPBlock([512, 256]))
+      two_tower = ml.ParallelBlock({"user": user_tower, "item": item_tower})
+      model = ml.Model(two_tower, ml.ItemRetrievalTask())
+      optimizer = MultiOptimizer(default_optimizer="adam",
+        optimizers_and_blocks=[
+          (tf.keras.optimizers.SGD(), user_tower),
+          (tf.keras.optimizers.Adam(), item_tower),
+        ])
+
+      # The string identification of optimizer is also acceptable:
+      optimizer = MultiOptimizer(default_optimizer="adam",
+        optimizers_and_blocks=[
+          ("sgd", user_tower),
+          ("adam", item_tower),
+        ])
+    ```
+
+    """
+
+    def __init__(
+        self,
+        default_optimizer: tf.keras.optimizers.Optimizer,
```

Comment:
Shouldn’t this be optional, since it could be that the user provides a optimizer for all variables using the `optimizers_and_blocks` argument?

Reply:
Yeah, make sense, I think we can make it optional, and if there is variable not given in optimizers_and_blocks by users, and also do not set this default_optimizer argument, we could warn that some variables would be assigned with the default default_optimizer.